### PR TITLE
global: change e-mails to notifications

### PR DESCRIPTION
**Pull request by ntarocco**
- implements a more generic way to send notifications using multiple
  backends. By default it will send e-mails.
- adds the possibility to filter what notifications to send depending on
  the record and the action that triggered the notifications.

### CODE REVIEW

**Comment by ntarocco** [comment]
@kprzerwa ths PR adds the new code for notifications
## TODO

### Here
- replace the various `send_*_mail` functions with the `send_*_notifications` functions for the main module, circulation, document requests and ILL
- fix tests (there are still some tests about mails in the other modules)
- test each action to make sure e-mails are sent correctly
- remove all code about e-mails, old config vars

### React app:
- replace all e-mails mention/cmps with notifications. Use the new endpoints `/notification` (payload did not change)

### CDS-ILS
- copy/paste the mail templates in a new folder `notifications` and replace the `subject` block with `title`
- override/redefine the config vars, in particular the notifications filter functions and disabled e-mails that should not be sent.

File: invenio_app_ils/config.py [hunk 1]
```diff
@@ -148,6 +148,30 @@ def _(x):
...
+###############################################################################
+# The backends to use when sending notifications.
+ILS_NOTIFICATIONS_BACKENDS_BUILDER = "invenio_app_ils.notifications.backends:notifications_backend_builder"
+# Notification message creator
+ILS_NOTIFICATIONS_MSG_BUILDER = "invenio_app_ils.notifications.messages:notification_msg_builder"  # noqa
+# Override default global common templates
+ILS_NOTIFICATIONS_TEMPLATES = {"footer": "footer.html"}
+# CIRCULATION
+# Notification message creator for loan notifications
+ILS_NOTIFICATIONS_MSG_BUILDER_CIRCULATION = "invenio_app_ils.circulation.notifications.messages:notification_loan_msg_builder"  # noqa
```
Comment (ntarocco) [inline invenio_app_ils/config.py]: circulation configs should be probably moved to `invenio_app_ils/circulation/config.py`

File: docker-services.yml [hunk 2]
```diff
@@ -52,8 +52,8 @@ services:
   db:
     image: postgres:12.4
     environment:
-      - "POSTGRES_USER=test"
-      - "POSTGRES_PASSWORD=psw"
+      - "POSTGRES_USER=invenio"
```
Comment (ntarocco) [inline docker-services.yml]: This should not probably be committed? The default config in config.py is:
`SQLALCHEMY_DATABASE_URI = "postgresql+psycopg2://test:psw@localhost/ils"`

File: invenio_app_ils/circulation/notifications/api.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+def send_expiring_loan_reminder_notification(loan, expiring_in_days):
+    """Send reminder email."""
+    send_loan_notification(
+        action="expiring_reminder",
+        loan=loan,
+        msg_extra_ctx=dict(expiring_in_days=expiring_in_days),
+    )
+
+
+def circulation_filter_notifications(*args, **kwargs):
```
Comment (ntarocco) [inline invenio_app_ils/circulation/notifications/api.py]:
```suggestion
def circulation_filter_notifications(loan, action, **kwargs):
```

File: invenio_app_ils/config.py [hunk 4]
```diff
@@ -252,7 +272,7 @@ def _(x):
 # Database
 ###############################################################################
 #: Database URI including user and password
-SQLALCHEMY_DATABASE_URI = "postgresql+psycopg2://test:psw@localhost/ils"
+SQLALCHEMY_DATABASE_URI = "postgresql+psycopg2://invenio:invenio@localhost/ils"
```
Comment (ntarocco) [inline invenio_app_ils/config.py]: to be reverted?

File: invenio_app_ils/document_requests/notifications/api.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    builder = obj_or_import_string(func_or_path)
+    msg = builder(document_request, action, msg_ctx, **kwargs)
+
+    patron = msg_ctx["patron"]
+    patrons = [patron]
+
+    send_notification(patrons, msg, **kwargs)
+
+
+def document_request_notification_filter(*args, **kwargs):
```
Comment (ntarocco) [inline invenio_app_ils/document_requests/notifications/api.py]:
```suggestion
def document_request_notification_filter(document_request, action, **kwargs):
```

File: invenio_app_ils/ill/notifications/api.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    builder = obj_or_import_string(func_or_path)
+    msg = builder(record=brw_req, action=action, msg_ctx=msg_ctx, **kwargs)
+
+    patron = msg_ctx["patron"]
+    patrons = [patron]
+
+    send_notification(patrons, msg, **kwargs)
+
+
+def ill_notifications_filter(*args, **kwargs):
```
Comment (ntarocco) [inline invenio_app_ils/ill/notifications/api.py]:
```suggestion
def ill_notifications_filter(brw_req, action, **kwargs):
```

File: tests/api/conftest.py [hunk 7]
```diff
@@ -56,6 +53,9 @@ def app_config(app_config):
         "RATELIMIT_AUTHENTICATED_USER": "1000 per minute",
         "CIRCULATION_TRANSACTION_USER_VALIDATOR": lambda x: True,
         "EXTEND_LOANS_LOCATION_UPDATED": False,
+        "SQLALCHEMY_DATABASE_URI":
+            "postgresql+psycopg2://invenio:invenio@localhost/ils",
```
Comment (ntarocco) [inline tests/api/conftest.py]: This is not needed normally, it is already set [here](https://github.com/inveniosoftware/docker-services-cli/blob/347bce92f8abfc07b58e54850d8a38cf1c89596d/docker_services_cli/config.py#L57) to use the test docker services [here](https://github.com/inveniosoftware/docker-services-cli/blob/347bce92f8abfc07b58e54850d8a38cf1c89596d/docker_services_cli/config.py#L52).

File: tests/api/ils/test_notifications_mails.py [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+    fake_recipients = app_with_notifs.config["ILS_MAIL_NOTIFY_TEST_RECIPIENTS"]
+    with app_with_notifs.extensions["mail"].record_messages() as outbox:
+        assert len(outbox) == 0
+        send_notification([patron], msg)
+        assert succ.s.called
+        assert len(outbox) == 1
+        assert outbox[0].recipients == fake_recipients
+
+    app_with_notifs.config["ILS_MAIL_ENABLE_TEST_RECIPIENTS"] = False
```
Comment (ntarocco) [inline tests/api/ils/test_notifications_mails.py:39]: can we add a test to ensure that the new functions that filter what notifications to send are working properly?

File: invenio_app_ils/config.py (hunk 4 above)
Comment (kpsherva) [reply invenio_app_ils/config.py]: I cannot run it with docker-services-cli if I don't configure it in this way, I had the same problem with cds-ils

Comment (ntarocco) [reply invenio_app_ils/config.py]: let's see this together, I am worried that after merging this everyone else will be impacted.

**Comment by ntarocco** [comment]
I also found other mail stuff searching by `mail` lowercase:
* `email-overdue` endpoint
* `ILS_CIRCULATION_MAIL_TEMPLATES` and similar
* `ILS_ILL_MAIL_MSG_CREATOR`
* `send_email_after_loan_change`

**Pull request merged**